### PR TITLE
#164798927 Implementation of role based access control

### PR DESCRIPTION
**Pull request by TheDrizzyWay**
## Description
This PR adds a role based access functionality which restricts users actions, based on their role and permission, and enables superadmins change the roles of other users..

## Type of change
- New feature

Screenshot 2019-04-04 at 5 50 00 PM
Screenshot 2019-04-03 at 6 07 30 PM

## Checklist:
- [x]  Write unit tests for this feature
- [x]  Set up middleware for authorizing users.
- [x] Create controller for updating user roles.
- [x] Add API documentation using swagger

## PIVOTAL BOARD
[#164798927](https://www.pivotaltracker.com/story/show/164798927)

### CODE REVIEW

File: server/controllers/user.controller.js [hunk 1]
```diff
@@ -1,10 +1,8 @@
 import dotenv from 'dotenv';
 import db from '../models';
-import utils from '../utils';
-import sendMail from '../utils/mail.util';
+import { passwordHash, generateToken, Response, sendMail } from '../utils';
```
Comment (chukwudidu) [inline server/controllers/user.controller.js]: Thanks for refactoring this.

File: server/routes/auth.routes.js [hunk 2]
```diff
@@ -1,13 +1,10 @@
 import express, { Router } from 'express';
 import passport from 'passport';
 import controllers from '../controllers';
-import utils from '../utils';
+import { tryCatch, userSignup, userEmpty, userExist, isSigninFieldEmpty } from '../utils';
```
Comment (chukwudidu) [inline server/routes/auth.routes.js]: Nice

**Comment by chukwudidu** [comment]
Looks good.

File: server/migrations/20190325190151-create-user.js [hunk 3]
```diff
@@ -24,8 +24,13 @@ module.exports = {
         type: Sequelize.STRING,
         allowNull: false
       },
+      role: {
+        type: Sequelize.ENUM('superadmin', 'admin', 'user'),
+        allowNull: false,
+        defaultValue: 'user'
+      },
```
Comment (samorahno) [inline server/migrations/20190325190151-create-user.js:31]: Good.

File: server/utils/users/permissions.util.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    return Response(res, 403, 'Access Denied. For Superadmins only.');
+  }
+  return next();
+};
+
+const adminAuth = (req, res, next) => {
+  const { role } = req.user;
+  if (role !== 'superadmin' && role !== 'admin') {
+    return Response(res, 403, 'Access Denied. You do not have the permission to perform this action.');
+  }
```
Comment (samorahno) [inline server/utils/users/permissions.util.js:56]: Good job

File: server/utils/users/permissions.util.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  if (!token) {
+    return Response(res, 401, 'Access Denied. Please Log In.');
+  }
+
+  try {
+    const decoded = await verifyToken(token);
+    req.user = decoded;
+    next();
+  } catch (error) {
+    return Response(res, 500, 'Error in verification. Please try again');
```
Comment (davidokonji) [inline server/utils/users/permissions.util.js]: @TheDrizzyWay  Hey, error code 500  is for server errors, for verification should be either 401 or 403 which ever suites the purpose of the response

File: server/controllers/user.controller.js [hunk 6]
```diff
@@ -144,6 +143,36 @@ class UserController {
...
     }
   }
+
+  /**
+     * @description updates the role of a user
+     * @param {object} req
+     * @param {object} res
+     * @returns {object} User Profile Object
+     */
+  static async updateUserRole(req, res) {
```
Comment (davidokonji) [inline server/controllers/user.controller.js:158]: Good work

File: server/utils/users/permissions.util.js (hunk 5 above)
Comment (TheDrizzyWay) [reply server/utils/users/permissions.util.js]: @davidokonji Noted. will fix right away

File: server/controllers/user.controller.js [hunk 7]
```diff
@@ -144,6 +143,36 @@ class UserController {
...
+    const { id } = req.params;
+    const foundUser = await User.findByPk(id);
+
+    if (!foundUser) return Response(res, 404, 'This user was not found.');
+
+    const newRole = foundUser.role === 'user' ? 'admin' : 'user';
```
Comment (paradisekelechi) [inline server/controllers/user.controller.js]: This just changes the role to `user` if it was `admin` and vice versa. 
Why not state the role you want when updating as coming from the request body? 

This would necessitate a `roles` table that would be associated to the user table. This would also make it possible to plug in more roles into the application as it scales

File: server/controllers/user.controller.js [hunk 8]
```diff
@@ -144,6 +143,36 @@ class UserController {
...
+    const updatedUser = await User.update({ role: newRole }, {
+      returning: true,
+      where: { id },
+    });
+
+    const updatedResult = updatedUser[1][0].dataValues;
+    const result = {
+      id: updatedResult.id,
```
Comment (paradisekelechi) [inline server/controllers/user.controller.js]: You can properly destructure the `updatedResult` object

File: server/models/user.js [hunk 9]
```diff
@@ -25,8 +25,13 @@ module.exports = (sequelize, DataTypes) => {
...
       allowNull: false
     },
+    role: {
+      type: DataTypes.ENUM('superadmin', 'admin', 'user'),
+      allowNull: false,
+      defaultValue: 'user'
+    },
     social: {
-      type: DataTypes.ENUM('facebook', 'twitter', 'google', null),
+      type: DataTypes.STRING,
```
Comment (paradisekelechi) [inline server/models/user.js:34]: Why this change to `STRING` and how does it affect the role based access control feature?

**Review by paradisekelechi — CHANGES_REQUESTED** [review]

File: server/controllers/user.controller.js (hunk 7 above)
Comment (TheDrizzyWay) [reply server/controllers/user.controller.js]: Got it. Will fix that.

File: server/models/user.js (hunk 9 above)
Comment (TheDrizzyWay) [reply server/models/user.js:34]: It seems sequelize has an issue with having two ENUM columns on the same table, hence the change

File: server/controllers/user.controller.js [hunk 10]
```diff
@@ -144,6 +143,37 @@ class UserController {
...
+
+    if (!foundUser) return Response(res, 404, 'This user was not found.');
+
+    const newRoleId = await Role.findOne({ where: { type: newrole } });
+    const updatedUser = await User.update({ roleId: newRoleId.dataValues.id }, {
+      returning: true,
+      where: { id },
+    });
+
+    const { dataValues } = updatedUser[1][0];
```
Comment (paradisekelechi) [inline server/controllers/user.controller.js]: Destructure dataValues further. You would have something like this: 
```
const { dataValues: { id, firstname, lastname, newrole, updatedAt } } = updatedUser[1][0];
const result = {
  id, 
  firstname,
  lastname,
  role: newrole,
  updatedAt
}


Implement this here and in every other place necessary

**Review by geekdanitek — APPROVED** [review]
Looks good.

File: server/controllers/user.controller.js (hunk 10 above)
Comment (TheDrizzyWay) [reply server/controllers/user.controller.js]: Fixed it. Thanks for the example.

File: server/controllers/user.controller.js [hunk 11]
```diff
@@ -144,6 +143,37 @@ class UserController {
...
+
+    if (!foundUser) return Response(res, 404, 'This user was not found.');
+
+    const newRoleId = await Role.findOne({ where: { type: newrole } });
+    const updatedUser = await User.update({ roleId: newRoleId.dataValues.id }, {
+      returning: true,
+      where: { id: userId },
+    });
+
+    const { dataValues: { id, firstname, lastname, updatedAt } } = updatedUser[1][0];
```
[the remark this answers is not in the document]
Comment (paradisekelechi) [reply server/controllers/user.controller.js]: @TheDrizzyWay Why did you resolve this hound issue?

[the remark this answers is not in the document]
Comment (TheDrizzyWay) [reply server/controllers/user.controller.js]: @paradisekelechi That was an error. I have unresolved it and pushed changes

File: server/controllers/user.controller.js [hunk 12]
```diff
@@ -144,6 +148,44 @@ class UserController {
...
+     * @returns {object} User Profile Object
+     */
+  static async updateUserRole(req, res) {
+    const userId = req.params.id;
+    const { newrole } = req.body;
+    const foundUser = await User.findByPk(userId);
+
+    if (!foundUser) return Response(res, 404, 'This user was not found.');
+
+    const newRoleId = await Role.findOne({ where: { type: newrole } });
```
Comment (paradisekelechi) [inline server/controllers/user.controller.js:165]: @TheDrizzyWay What happens when the role is not found in the Role table?

Comment (paradisekelechi) [reply server/controllers/user.controller.js:165]: More so, it makes more sense to search for the role by the id or uuid.

Comment (TheDrizzyWay) [reply server/controllers/user.controller.js:165]: @paradisekelechi there's a middleware that throws an error, when the role entered does not exist. The request will not reach the database. For the part of searching via uuid, do i still use req.body for that or do i use req.params instead?

Comment (TheDrizzyWay) [reply server/controllers/user.controller.js:165]: @paradisekelechi changes implemented and screenshot updated. please review

**Review by paradisekelechi — APPROVED** [review]
